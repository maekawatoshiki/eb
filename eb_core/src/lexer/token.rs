use super::{location::Location, source::Source};
use nom::{
    branch::alt,
    bytes::complete::{tag, take_until, take_while1},
    character::complete::{char, digit1, multispace0},
    combinator::map,
    error::VerboseError,
    multi::many1,
    sequence::{preceded, terminated, tuple},
    IResult,
};

#[derive(Debug, Clone, PartialEq)]
pub struct Token<'a> {
    kind: TokenKind<'a>,
    loc: Location,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenKind<'a> {
    Int(&'a str),
    Ident(&'a str),
    OpenDelim(DelimKind),
    CloseDelim(DelimKind),
    Punct(PunctKind),
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum DelimKind {
    Paren,
    Bracket,
    Brace,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum PunctKind {
    Plus,
    Minus,
    Star,
    Slash,
    Eq,
    Colon,
    DoubleSemicolon,
    Comma,
}

pub struct TokenStream<'a> {
    source: &'a Source,
    tokens: Vec<Token<'a>>,
    body: &'a String,
    cur: &'a str,
}

impl<'a> Token<'a> {
    pub fn new(kind: TokenKind<'a>, loc: Location) -> Self {
        Self { kind, loc }
    }

    pub fn kind(&self) -> &TokenKind {
        &self.kind
    }

    pub fn loc(&self) -> &Location {
        &self.loc
    }
}

impl<'a> TokenKind<'a> {
    pub fn as_ident(&self) -> Option<&'a str> {
        match self {
            Self::Ident(i) => Some(i),
            _ => None,
        }
    }

    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            ":" => Some(Self::Punct(PunctKind::Colon)),
            ";;" => Some(Self::Punct(PunctKind::DoubleSemicolon)),
            "," => Some(Self::Punct(PunctKind::Comma)),
            "(" => Some(Self::OpenDelim(DelimKind::Paren)),
            ")" => Some(Self::CloseDelim(DelimKind::Paren)),
            "{" => Some(Self::OpenDelim(DelimKind::Brace)),
            "}" => Some(Self::CloseDelim(DelimKind::Brace)),
            "[" => Some(Self::OpenDelim(DelimKind::Bracket)),
            "]" => Some(Self::CloseDelim(DelimKind::Bracket)),
            "+" => Some(Self::Punct(PunctKind::Plus)),
            "-" => Some(Self::Punct(PunctKind::Minus)),
            "*" => Some(Self::Punct(PunctKind::Star)),
            "/" => Some(Self::Punct(PunctKind::Slash)),
            "==" => Some(Self::Punct(PunctKind::Eq)),
            _ => None,
        }
    }
}

impl<'a> TokenStream<'a> {
    pub fn new(source: &'a Source) -> Self {
        let body = source.body();
        Self {
            source,
            tokens: vec![],
            body,
            cur: body.as_str(),
        }
    }

    pub fn source(&self) -> &Source {
        self.source
    }
}

impl<'a> Iterator for TokenStream<'a> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        let bgn = self.body.as_str().as_ptr() as usize;
        let loc = |source: &str| -> Location { Location((source.as_ptr() as usize - bgn) as u32) };
        if let Some((source, token)) = preceded(
            spaces,
            alt((
                map(digit1, |i: &str| Token::new(TokenKind::Int(i), loc(i))),
                map(delimiter, |s: &str| {
                    Token::new(TokenKind::from_str(s).unwrap(), loc(s))
                }),
                map(symbol, |s: &str| {
                    Token::new(TokenKind::from_str(s).unwrap(), loc(s))
                }),
                map(identifier, |i: &str| {
                    Token::new(TokenKind::Ident(i), loc(i))
                }),
            )),
        )(self.cur)
        .ok()
        {
            self.cur = source;
            self.tokens.push(token.clone());
            return Some(token);
        }
        None
    }
}

pub fn symbol(source: &str) -> IResult<&str, &str, VerboseError<&str>> {
    alt((
        tag(":"),
        tag(";;"),
        tag(","),
        tag("+"),
        tag("-"),
        tag("*"),
        tag("/"),
        tag("=="),
    ))(source)
}

pub fn delimiter(source: &str) -> IResult<&str, &str, VerboseError<&str>> {
    alt((tag("("), tag(")"), tag("["), tag("]"), tag("{"), tag("}")))(source)
}

pub fn identifier(source: &str) -> IResult<&str, &str, VerboseError<&str>> {
    take_while1(|c: char| c.is_alphanumeric() || c == '_')(source)
}

pub fn spaces(source: &str) -> IResult<&str, (), VerboseError<&str>> {
    alt((
        map(
            many1(tuple((
                multispace0,
                preceded(tag("//"), terminated(take_until("\n"), char('\n'))),
                multispace0,
            ))),
            |_| (),
        ),
        map(multispace0, |_| ()),
    ))(source)
}
