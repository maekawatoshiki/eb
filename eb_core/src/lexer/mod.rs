pub mod location;
pub mod source;
pub mod token;

use source::Source;

pub fn tokenize<'a>(source: &'a Source) -> token::TokenStream<'a> {
    token::TokenStream::new(source)
}

#[test]
fn tokenize1() {
    use location::Location;
    use token::Token;

    let source = Source::String(r#"func f(x i32) i32: x;;"#.to_string());
    let tokenize: Vec<Token> = tokenize(&source).into_iter().collect();
    let correct = vec![
        Token::new(token::TokenKind::Ident("func"), Location(0)),
        Token::new(token::TokenKind::Ident("f"), Location(4)),
        Token::new(
            token::TokenKind::OpenDelim(token::DelimKind::Paren),
            Location(6),
        ),
        Token::new(token::TokenKind::Ident("x"), Location(7)),
        Token::new(token::TokenKind::Ident("i32"), Location(8)),
        Token::new(
            token::TokenKind::CloseDelim(token::DelimKind::Paren),
            Location(12),
        ),
        Token::new(token::TokenKind::Ident("i32"), Location(13)),
        Token::new(token::TokenKind::Colon, Location(17)),
        Token::new(token::TokenKind::Ident("x"), Location(18)),
        Token::new(token::TokenKind::DoubleSemicolon, Location(20)),
    ];
    assert_eq!(tokenize.len(), correct.len());
    assert!(tokenize.iter().zip(correct.iter()).all(|(a, b)| a == b))
}

#[test]
fn tokenize2() {
    use location::Location;
    use token::{BinOpKind, DelimKind, Token, TokenKind};

    let source = Source::String(
        r#"
        func f(x y i32) i32: 
            x + 1 == y - 1;;
        "#
        .to_string(),
    );
    let tokenize: Vec<Token> = tokenize(&source).into_iter().collect();
    for token in &tokenize {
        println!("{:?}", token);
    }
    let correct = vec![
        Token::new(TokenKind::Ident("func"), Location(0)),
        Token::new(TokenKind::Ident("f"), Location(13)),
        Token::new(TokenKind::OpenDelim(DelimKind::Paren), Location(15)),
        Token::new(TokenKind::Ident("x"), Location(16)),
        Token::new(TokenKind::Ident("y"), Location(17)),
        Token::new(TokenKind::Ident("i32"), Location(19)),
        Token::new(TokenKind::CloseDelim(DelimKind::Paren), Location(23)),
        Token::new(TokenKind::Ident("i32"), Location(24)),
        Token::new(TokenKind::Colon, Location(28)),
        Token::new(TokenKind::Ident("x"), Location(29)),
        Token::new(TokenKind::BinOp(BinOpKind::Plus), Location(44)),
        Token::new(TokenKind::Int("1"), Location(46)),
        Token::new(TokenKind::BinOp(BinOpKind::Eq), Location(48)),
        Token::new(TokenKind::Ident("y"), Location(51)),
        Token::new(TokenKind::BinOp(BinOpKind::Minus), Location(53)),
        Token::new(TokenKind::Int("1"), Location(55)),
        Token::new(TokenKind::DoubleSemicolon, Location(57)),
    ];
    assert_eq!(tokenize.len(), correct.len());
    assert!(tokenize.iter().zip(correct.iter()).all(|(a, b)| a == b))
}
