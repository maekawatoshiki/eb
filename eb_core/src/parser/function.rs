use super::{expr, Context};
use crate::{
    ast::{expr as ast_expr, function as ast_func},
    lexer::token::{DelimKind, PunctKind},
};
use anyhow::Result;

pub fn parse(ctx: &mut Context) -> Result<ast_func::Node> {
    let loc = ctx.cur_loc()?;
    ctx.expect_keyword("func")?;
    let ident = ctx
        .expect_any_ident()?
        .kind()
        .as_ident()
        .unwrap()
        .to_string();
    ctx.expect_open_delim(DelimKind::Paren)?;
    let params = parse_parameters(ctx)?;
    ctx.expect_punct(PunctKind::Colon)?;
    let body = parse_body(ctx)?;
    Ok(ast_func::Node::new(ident, params, body, loc))
}

fn parse_parameters(ctx: &mut Context) -> Result<Vec<ast_func::Param>> {
    if ctx.skip_close_delim(DelimKind::Paren) {
        return Ok(vec![]);
    }
    let mut params = vec![];

    loop {
        let param = ctx
            .expect_any_ident()?
            .kind()
            .as_ident()
            .unwrap()
            .to_string();
        params.push(ast_func::Param::new(param));

        if ctx.skip_punct(PunctKind::Comma) {
            continue;
        }

        ctx.expect_close_delim(DelimKind::Paren)?;

        return Ok(params);
    }
}

fn parse_body(ctx: &mut Context) -> Result<Vec<ast_expr::Node>> {
    if ctx.skip_punct(PunctKind::DoubleSemicolon) {
        return Ok(vec![]);
    }

    let mut body = vec![];

    loop {
        body.push(expr::parse(ctx)?);
        if ctx.skip_punct(PunctKind::DoubleSemicolon) {
            return Ok(body);
        }
    }
}

#[cfg(test)]
mod test {
    extern crate insta;
    use super::*;
    use crate::lexer::{source::Source, tokenize};

    #[test]
    fn parse1() {
        let source = Source::String(r#"func f(): ;;"#.to_string());
        let mut ctx = Context::new(tokenize(&source));
        insta::assert_debug_snapshot!(parse(&mut ctx).expect("fail to parse"));
    }

    #[test]
    fn parse2() {
        let source = Source::String(r#"func f(x): ;;"#.to_string());
        let mut ctx = Context::new(tokenize(&source));
        insta::assert_debug_snapshot!(parse(&mut ctx).expect("fail to parse"));
    }

    #[test]
    fn parse3() {
        let source = Source::String(r#"func f(x, y): ;;"#.to_string());
        let mut ctx = Context::new(tokenize(&source));
        insta::assert_debug_snapshot!(parse(&mut ctx).expect("fail to parse"));
    }

    #[test]
    fn parse4() {
        let source = Source::String(r#"func f(x): x;;"#.to_string());
        let mut ctx = Context::new(tokenize(&source));
        insta::assert_debug_snapshot!(parse(&mut ctx).expect("fail to parse"));
    }

    #[test]
    fn parse5() {
        let source = Source::String(r#"func f(x): x;;"#.to_string());
        let mut ctx = Context::new(tokenize(&source));
        insta::assert_debug_snapshot!(parse(&mut ctx).expect("fail to parse"));
    }
}
