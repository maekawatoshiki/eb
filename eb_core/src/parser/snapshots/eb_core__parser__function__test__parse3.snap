---
source: src/parser/function.rs
expression: "parse(&mut ctx).expect(\"fail to parse\")"

---
Node {
    name: "f",
    params: [
        Param {
            name: "x",
        },
        Param {
            name: "y",
        },
    ],
    body: [],
    loc: Location(
        0,
    ),
}
